File: fastq_stats.jl
#!/usr/bin/env julia

using Printf
using Statistics
using Klib

const partitions = [0.1, 0.15, 0.2, 0.25, 0.3, 0.35, 0.4, 0.45, 0.5, 0.55, 0.6, 0.65, 0.7, 0.75, 0.8, 0.85, 0.9]

function get_Nxx(seq_lengths::Array{Int,1}, slen::Int)
    thresholds = partitions * slen
    # Nxx_len = zeros(Int, (length(partitions), 3))
    Nxx_len = zeros(Int, (3, length(partitions)))
    accu_len = 0
    n = 0
    for i in 1:1:length(partitions)
        while accu_len < thresholds[i]
            n += 1
            accu_len += seq_lengths[n]
        end
        # Nxx_len[i,:] = [n, seq_lengths[n], accu_len]
        Nxx_len[:,i] = [n, seq_lengths[n], accu_len]
    end
    return Nxx_len
end

function generate_QualDict()
    # phred 33
    code = Vector{UInt8}("""!"#\$%&'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\\]^_`abcdefghijklmnopqrstuvwxyz{|}~""")
    @assert length(code) == 94
    d = Dict{UInt8, Float32}()
    for c in code
        d[c] = Float32(10) ^ ((c - UInt8(33)) / Float32(-10.0))
    end
    return d
end
const QD = generate_QualDict()

function cal_meadQ(qual::String)::Float32
    qualBytes = Vector{UInt8}(qual)
    sumProb = Float32(0)
    for i in qualBytes
        sumProb += QD[i]
    end
    avgProb = sumProb / Float32(length(qualBytes))
    Float32(-10.0) * log10(avgProb)
end

# function cal_meadQ(qual::String)::Float32
#     qualBytes = Vector{UInt8}(qual)
#     ## t1
#     # avgProb = sum(Float32(10.0) .^ ((qualBytes .- UInt8(33)) / Float32(-10.0))) / Float32(length(qualBytes))
#     ## t2
#     sumProb = Float32(0)
#     for i in qualBytes
#         sumProb += Float32(10) ^ ((i - UInt8(33)) / Float32(-10.0))
#     end
#     avgProb = sumProb / Float32(length(qualBytes))
#     Float32(-10.0) * log10(avgProb)
# end

function usage()
    println("Usage: fastq_stats.jl [-q] [-s] [--dump FILE] <in.fq.gz>")
    exit(1)
end

function main(args::Vector{String})
    ### handle opts ###

    readQ = true
    singleLine = false
    dumpToFile = false
    dumpArg = ""
    for (opt, arg) in Klib.getopt(args, "qs", ["dump="])
        if opt == "-q"
            readQ = true
        elseif opt == "-s"
            singleLine = true
        elseif opt == "--dump"
            dumpToFile = true
            dumpArg = arg
        else
            usage()
        end
    end
    if length(args) != 1
        usage()
    end

    if args[1] == "-"
        fn = stdin
    else
        fn = Klib.GzFile(args[1])
    end
    fx = Klib.FastxReader(fn)

    if dumpToFile
        if dumpArg == "-" || dumpArg == "auto"
            fnDump = open(args[1]*".dump", "w")
        else
            fnDump = open(dumpArg, "w")
        end
    end

    ### main ###

    n, slen = 0, 0
    seq_lengths = Int[]
    seq_meanQ = Float32[]
    while (r = read(fx)) != nothing
        n += 1
        push!(seq_lengths, sizeof(r.seq))
        slen += sizeof(r.seq)
        if readQ
            q = cal_meadQ(r.qual)
            push!(seq_meanQ, q)
            if dumpToFile
                write(fnDump, @sprintf "%s\t%s\t%.6f\n" r.name sizeof(r.seq) q)
            end
        end
    end
    if dumpToFile
        close(fnDump)
    end
    sort!(seq_lengths, rev=true)
    Nxx_len = get_Nxx(seq_lengths, slen)

    if singleLine
        @printf "%d\t%d\t%.1f\t%d\t%d\t%d\t%d\t\n" slen n slen/n seq_lengths[1] seq_lengths[end] Nxx_len[2,9] Nxx_len[2,17]
    else
        @printf "input file: %s\n" args[1]
        @printf "\n"
        @printf "minimum length: %d\n" seq_lengths[end]
        @printf "maximum length: %d\n" seq_lengths[1]
        @printf "total seqs:     %d\n" n
        @printf "total length:   %d (%.2f Gbp)\n" slen slen/1e9
        @printf "avg. length:    %.1f\n" slen/n
        @printf "median length:  %.1f\n" median(seq_lengths)
        @printf "N50:            %d\n" Nxx_len[2,9]
        @printf "N90:            %d\n" Nxx_len[2,17]
        @printf "\n"
        @printf "# total_base seq_num mean max min N50 N90\n"
        @printf "%d\t%d\t%.1f\t%d\t%d\t%d\t%d\t\n" slen n slen/n seq_lengths[1] seq_lengths[end] Nxx_len[2,9] Nxx_len[2,17]
        @printf "\n"
        @printf "# Nxx count Nxx_len accu_len\n"
        for (n,(a,b,c)) in zip(partitions, eachcol(Nxx_len))
            @printf "N%d\t%d\t%d\t%d\n" n*100 a b c
        end
        if readQ
            @printf "\n"
            @printf "# quality stats\n"
            @printf "Q>=20: %s\n" sum(seq_meanQ .>= 20)
            @printf "Q>=10: %s\n" sum(seq_meanQ .>= 10)
            @printf "Q>= 7: %s\n" sum(seq_meanQ .>= 7)
            @printf "Q>= 5: %s\n" sum(seq_meanQ .>= 5)
        end
    end

    # plot
    # TODO
end

main(ARGS)

File: fastq_stats.cpp
#include <zlib.h>
#include <unistd.h>
#include <math.h>
#include <stdio.h>
#include <stdlib.h>
#include <vector>
#include <algorithm>
#include <numeric>
#include <functional>
#include "kseq.h"
using namespace std;
KSEQ_INIT(gzFile, gzread)

// https://github.com/rrwick/Filtlong/blob/f0cb3b34907b9fdf204fda3a36f61cf07b50f68f/src/read.cpp#L270
// https://github.com/lh3/seqtk/blob/7c04ce7898ad5909bd309c6ba3cd9c3bd0651f0e/seqtk.c#L1843
// https://lellansin.wordpress.com/2015/05/18/cc-使用-memoization-优化算法/
inline float qscore_to_quality(char qscore, float *memo) {
    int q = qscore - 33;
    if (memo[q] == -1.0)
        memo[q] = powf(10.0, -q / 10.0);
    return memo[q];
}

float get_mean_quality(char *qscores, float *memo) {
    int length = strlen(qscores);
    float sum = 0.0;
    for (int i = 0; i < length; ++i) {
        sum += qscore_to_quality(qscores[i], memo);
    }
    return -10.0f * log10f(sum / length);
}

inline void print_seq(FILE *fp, const kseq_t *seq) {
    if (seq->qual.l) {
        // fastq
        fputc('@', fp); fputs(seq->name.s, fp);
        if (seq->comment.l)
            fputc(' ', fp); fputs(seq->comment.s, fp);
        fputc('\n', fp);
        fputs(seq->seq.s, fp); fputc('\n', fp);
        fputc('+', fp); fputc('\n', fp);
        fputs(seq->qual.s, fp); fputc('\n', fp);
    } else {
        // fasta
        fputc('>', fp); fputs(seq->name.s, fp);
        if (seq->comment.l)
            fputc(' ', fp); fputs(seq->comment.s, fp);
        fputc('\n', fp);
        fputs(seq->seq.s, fp); fputc('\n', fp);
    }
}

vector<unsigned long> get_Nxx(
        unsigned long &seqNumber,
        unsigned long long &totLen,
        vector<unsigned long> &seqLengths
        ) {
    // assume sorted seqLengths
    vector<float> partitions {0.5, 0.9}; // N50, N90
    vector<double> thresholds(partitions.size());
    vector<unsigned long> Nxx(partitions.size());

    for (unsigned int i = 0; i < partitions.size(); ++i) {
        thresholds[i] = totLen * partitions[i];
    }

    unsigned long long sum = 0;
    unsigned int j = 0;
    for (unsigned int i = 0; i < seqNumber; ++i) {
        sum += seqLengths[i];
        if (sum > thresholds[j]) {
            Nxx[j] = seqLengths[i];
            ++j;
            if (j == partitions.size())
                break;
        }
    }
    return Nxx;
}

struct ReadStats {
    // length stat, count, base
    unsigned long lc500;
    unsigned long lc1000;
    unsigned long lc5000;
    unsigned long lc10000;
    unsigned long lc30000;
    unsigned long long lb500;
    unsigned long long lb1000;
    unsigned long long lb5000;
    unsigned long long lb10000;
    unsigned long long lb30000;
    // quality stat, count, base
    unsigned long qc5;
    unsigned long qc7;
    unsigned long qc10;
    unsigned long qc15;
    unsigned long qc20;
    unsigned long long qb5;
    unsigned long long qb7;
    unsigned long long qb10;
    unsigned long long qb15;
    unsigned long long qb20;
};

ReadStats get_ReadStats(
        unsigned long &seqNumber,
        unsigned long long &totLen,
        vector<unsigned long> &seqLengths,
        vector<float> &seqQuals
        ) {
    // assume original order
    bool hasQ = (seqLengths.size() != seqQuals.size()) ? false : true;

    ReadStats out;
    out.lc500 = out.lc1000 = out.lc5000 = out.lc10000 = out.lc30000 = 0;
    out.lb500 = out.lb1000 = out.lb5000 = out.lb10000 = out.lb30000 = 0;
    out.qc5 = out.qc7 = out.qc10 = out.qc15 = out.qc20 = 0;
    out.qb5 = out.qb7 = out.qb10 = out.qb15 = out.qb20 = 0;

    for (unsigned int i = 0; i < seqNumber; ++i) {
        if (seqLengths[i] >= 500)
            ++out.lc500, out.lb500 += seqLengths[i];
        if (seqLengths[i] >= 1000)
            ++out.lc1000, out.lb1000 += seqLengths[i];
        if (seqLengths[i] >= 5000)
            ++out.lc5000, out.lb5000 += seqLengths[i];
        if (seqLengths[i] >= 10000)
            ++out.lc10000, out.lb10000 += seqLengths[i];
        if (seqLengths[i] >= 30000)
            ++out.lc30000, out.lb30000 += seqLengths[i];
        if (hasQ) {
            if (seqQuals[i] >= 5.0f)
                ++out.qc5, out.qb5 += seqLengths[i];
            if (seqQuals[i] >= 7.0f)
                ++out.qc7, out.qb7 += seqLengths[i];
            if (seqQuals[i] >= 10.0f)
                ++out.qc10, out.qb10 += seqLengths[i];
            if (seqQuals[i] >= 15.0f)
                ++out.qc15, out.qb15 += seqLengths[i];
            if (seqQuals[i] >= 20.0f)
                ++out.qc20, out.qb20 += seqLengths[i];
        }
    }
    return out;
}

template <typename T> 
double median(vector<T> &vec, int size, bool is_sorted=true) {
    if (is_sorted) {
        // works for both ascending and descending
        if (size % 2 == 0)
            return (vec[size/2-1] + vec[size/2]) / 2.0;
        else 
            return vec[size/2];
    } else {
        // original vector will be modified
        // O(n)
        if (size % 2 == 0) {
            T m1, m2;
            nth_element(vec.begin(), vec.begin()+(size/2-1), vec.end());
            m1 = vec[size/2-1];
            m2 = *min_element(vec.begin()+(size/2), vec.end());
            return (m1 + m2) / 2.0;
        } else {
            nth_element(vec.begin(), vec.begin()+(size/2), vec.end());
            return vec[size/2];
        }
    }
}

void generate_output(
        FILE *fp,
        unsigned long &seqNumber,
        unsigned long long &totLen,
        vector<unsigned long> &seqLengths,
        vector<float> &seqQuals
        ) {
    bool hasQ = (seqLengths.size() != seqQuals.size()) ? false : true;

    double totLenGb = 0, avgLen = 0, medLen = 0;
    totLenGb = totLen / 1000000000.0;
    avgLen = totLen / (double)seqNumber;
    ReadStats Stat = get_ReadStats(seqNumber, totLen, seqLengths, seqQuals);

    // sort here
    sort(seqLengths.begin(), seqLengths.end(), greater<>());
    auto Nxx = get_Nxx(seqNumber, totLen, seqLengths);
    medLen = median(seqLengths, seqNumber);

    double totQual = 0, avgQual = 0, medQual = 0;
    if (hasQ) {
        totQual = accumulate(seqQuals.begin(), seqQuals.end(), 0.0);
        avgQual = totQual / (double)seqNumber;
        medQual = median(seqQuals, seqNumber, false);
    }

    fprintf(fp, "minimum length: %lu\n", seqLengths.back());
    fprintf(fp, "maximum length: %lu\n", seqLengths.front());
    fprintf(fp, "total seqs:     %lu\n", seqNumber);
    fprintf(fp, "total length:   %llu (%.2f Gb)\n", totLen, totLenGb);
    fprintf(fp, "N50:            %lu\n", Nxx[0]);
    fprintf(fp, "N90:            %lu\n", Nxx[1]);
    fprintf(fp, "avg. length:    %.2f\n", avgLen);
    fprintf(fp, "median length:  %.2f\n", medLen);
    if (hasQ) {
    fprintf(fp, "avg. quality:   %.2f\n", avgQual);
    fprintf(fp, "median quality: %.2f\n", medQual);
    }
    fprintf(fp, "\n");
    fprintf(fp, "L>=30000: %lu (%.2f%%)\t%llu (%.2f%%)\n", Stat.lc30000, 100.0*Stat.lc30000/seqNumber, Stat.lb30000, 100.0*Stat.lb30000/totLen);
    fprintf(fp, "L>=10000: %lu (%.2f%%)\t%llu (%.2f%%)\n", Stat.lc10000, 100.0*Stat.lc10000/seqNumber, Stat.lb10000, 100.0*Stat.lb10000/totLen);
    fprintf(fp, "L>=5000 : %lu (%.2f%%)\t%llu (%.2f%%)\n", Stat.lc5000, 100.0*Stat.lc5000/seqNumber, Stat.lb5000, 100.0*Stat.lb5000/totLen);
    fprintf(fp, "L>=1000 : %lu (%.2f%%)\t%llu (%.2f%%)\n", Stat.lc1000, 100.0*Stat.lc1000/seqNumber, Stat.lb1000, 100.0*Stat.lb1000/totLen);
    fprintf(fp, "L>=500  : %lu (%.2f%%)\t%llu (%.2f%%)\n", Stat.lc500, 100.0*Stat.lc500/seqNumber, Stat.lb500, 100.0*Stat.lb500/totLen);
    if (hasQ) {
    fprintf(fp, "\n");
    fprintf(fp, "Q>=20: %lu (%.2f%%)\t%llu (%.2f%%)\n", Stat.qc20, 100.0*Stat.qc20/seqNumber, Stat.qb20, 100.0*Stat.qb20/totLen);
    fprintf(fp, "Q>=15: %lu (%.2f%%)\t%llu (%.2f%%)\n", Stat.qc15, 100.0*Stat.qc15/seqNumber, Stat.qb15, 100.0*Stat.qb15/totLen);
    fprintf(fp, "Q>=10: %lu (%.2f%%)\t%llu (%.2f%%)\n", Stat.qc10, 100.0*Stat.qc10/seqNumber, Stat.qb10, 100.0*Stat.qb10/totLen);
    fprintf(fp, "Q>=7 : %lu (%.2f%%)\t%llu (%.2f%%)\n", Stat.qc7, 100.0*Stat.qc7/seqNumber, Stat.qb7, 100.0*Stat.qb7/totLen);
    fprintf(fp, "Q>=5 : %lu (%.2f%%)\t%llu (%.2f%%)\n", Stat.qc5, 100.0*Stat.qc5/seqNumber, Stat.qb5, 100.0*Stat.qb5/totLen);
    }
}

void generate_output_short(
        FILE *fp,
        unsigned long &seqNumber,
        unsigned long long &totLen,
        vector<unsigned long> &seqLengths,
        vector<float> &seqQuals
        ) {
    bool hasQ = (seqLengths.size() != seqQuals.size()) ? false : true;

    double totLenGb = 0, avgLen = 0, medLen = 0;
    // totLenGb = totLen / 1000000000.0;
    avgLen = totLen / (double)seqNumber;
    // ReadStats Stat = get_ReadStats(seqNumber, totLen, seqLengths, seqQuals);

    // sort here
    sort(seqLengths.begin(), seqLengths.end(), greater<>());
    auto Nxx = get_Nxx(seqNumber, totLen, seqLengths);
    medLen = median(seqLengths, seqNumber);

    double totQual = 0, avgQual = 0, medQual = 0;
    if (hasQ) {
        totQual = accumulate(seqQuals.begin(), seqQuals.end(), 0.0);
        avgQual = totQual / (double)seqNumber;
        medQual = median(seqQuals, seqNumber, false);
    }

    if (hasQ) {
        // total_base seq_num max min N50 N90 mean median meanQ medianQ
        fprintf(fp, "%llu\t%lu\t%lu\t%lu\t%lu\t%lu\t%.2f\t%.2f\t%.2f\t%.2f\n", totLen, seqNumber, seqLengths.front(), seqLengths.back(), Nxx[0], Nxx[1], avgLen, medLen, avgQual, medQual);
    } else {
        // total_base seq_num max min N50 N90 mean median
        fprintf(fp, "%llu\t%lu\t%lu\t%lu\t%lu\t%lu\t%.2f\t%.2f\n", totLen, seqNumber, seqLengths.front(), seqLengths.back(), Nxx[0], Nxx[1], avgLen, medLen);
    }
}

void generate_output_diff(
        FILE *fp,
        unsigned long nB, unsigned long n,
        unsigned long long totLenB, unsigned long long totLen
        ) {
    fprintf(fp, "\n");
    fprintf(fp, "%lu/%lu (%.2f%%) reads, ", n, nB, (double)n/nB*100.0);
    fprintf(fp, "%llu/%llu (%.2f%%) bases are preserved.\n", totLen, totLenB, (double)totLen/totLenB*100.0);
}

int main(int argc, char *argv[]) {
	gzFile fp;
	kseq_t *seq;
    FILE *fpDump = NULL, *fpStatBefore = NULL;
    unsigned long filterLen;
    float filterQual;
    bool ignoreQ = false, shortOutput = false, doFilter = false;
    int c;
    while ((c = getopt(argc, argv, "Al:q:B:D:s")) >= 0) {
        switch (c) {
            case 'A': ignoreQ = true; break;
            case 'l': doFilter = true; filterLen = atoi(optarg); break;
            case 'q': doFilter = true; filterQual = atof(optarg); break;
            case 'B': fpStatBefore = fopen(optarg, "w"); break;
            case 'D': fpDump = fopen(optarg, "w"); break;
            case 's': shortOutput = true; break;
        }
    }

	if (argc == optind) {
		printf("Usage: fastq_stats [options] <FASTX>\n");
        printf("Options:\n");
        printf("    -A       ignore qualities\n");
        printf("    -l NUM   minimum length\n");
        printf("    -q NUM   minimum mean quality\n");
        printf("    -B FILE  print read stats before filtering\n");
        printf("    -D FILE  dump per read stats to file\n");
        printf("    -s       short output\n");
		return 1;
	}
    // https://github.com/lh3/seqtk/blob/7c04ce7898ad5909bd309c6ba3cd9c3bd0651f0e/seqtk.c#L1410
    fp = optind < argc && strcmp(argv[optind], "-") ? gzopen(argv[optind], "r") : gzdopen(fileno(stdin), "r");
	if (fp == 0) {
		fprintf(stderr, "Failed to open the input file/stream.\n");
		return 1;
	}
	seq = kseq_init(fp);

    // memoization
    float memo[100];
    for (int j = 0; j < 100; ++j)
        memo[j] = -1.0;

    // stats
	int r = 0;
    unsigned long n = 0;
    unsigned long long totLen = 0;
    vector<unsigned long> seqLengths;
    vector<float> seqQuals;
    float meanQ = 0.0;
    // stats before filtering
    unsigned long nB = 0;
    unsigned long long totLenB = 0;
    vector<unsigned long> seqLengthsB;
    vector<float> seqQualsB;
    // read seq
    if (doFilter) {
        while ((r = kseq_read(seq)) >= 0) {
            ++nB;
            totLenB += seq->seq.l;
            seqLengthsB.push_back(seq->seq.l);
            if (! ignoreQ && seq->qual.l) {
                meanQ = get_mean_quality(seq->qual.s, memo);
                seqQualsB.push_back(meanQ);
            }
            if (seq->seq.l >= filterLen && meanQ >= filterQual) {
                // pass
                print_seq(stdout, seq);
                ++n;
                totLen += seq->seq.l;
                seqLengths.push_back(seq->seq.l);
                if (! ignoreQ && seq->qual.l) {
                    seqQuals.push_back(meanQ);
                }
            }
            if (fpDump)
                fprintf(fpDump, "%s\t%lu\t%.6f\n", seq->name.s, seq->seq.l, meanQ);
        }
    } else {
        while ((r = kseq_read(seq)) >= 0) {
            ++n;
            totLen += seq->seq.l;
            seqLengths.push_back(seq->seq.l);
            if (! ignoreQ && seq->qual.l) {
                meanQ = get_mean_quality(seq->qual.s, memo);
                seqQuals.push_back(meanQ);
            }
            if (fpDump)
                fprintf(fpDump, "%s\t%lu\t%.6f\n", seq->name.s, seq->seq.l, meanQ);
        }
    }

	if (r != -1) fprintf(stderr, "ERROR: malformated FASTX\n");
	kseq_destroy(seq);
	gzclose(fp);
    if (fpDump) fclose(fpDump);

    // output stats
    // vectors might be modified and no longer preserve original order
    if (doFilter && fpStatBefore) {
        generate_output(fpStatBefore, nB, totLenB, seqLengthsB, seqQualsB);
        fclose(fpStatBefore);
    }

    if (shortOutput)
        generate_output_short(stderr, n, totLen, seqLengths, seqQuals);
    else
        generate_output(stderr, n, totLen, seqLengths, seqQuals);

    if (doFilter)
        generate_output_diff(stderr, nB, n, totLenB, totLen);

	return 0;
}
